Dasel query crashes with go panic

**Describe the bug**
dasel crashes with `panic: reflect.Set: value of type []interface {} is not assignable to type []*dencoding.Map`.
Maybe a dupe of #282.

<details>
<summary>Stacktrace</summary>

```
panic: reflect.Set: value of type []interface {} is not assignable to type []*dencoding.Map

goroutine 1 [running]:
reflect.Value.assignTo({0x7483e0?, 0xc0353721e0?, 0xc02880af88?}, {0x7b95e2, 0xb}, 0x7476a0, 0x0)
	reflect/value.go:3307 +0x289
reflect.Value.Set({0x7476a0?, 0xc028351f80?, 0x16?}, {0x7483e0?, 0xc0353721e0?, 0xc0285af860?})
	reflect/value.go:2260 +0xe6
github.com/tomwright/dasel/v2.Value.Append({{0xc0322d2340, 0xc028351f80, 0x16}, 0xc028706520, 0x0, 0xc0285af860})
	github.com/tomwright/dasel/v2/value.go:295 +0x285
github.com/tomwright/dasel/v2.glob..func3(0x760ce0?, 0x41679a?, {0xd5cfc0?, 0x6?, 0xa61cc0?})
	github.com/tomwright/dasel/v2/func_append.go:26 +0x2ae
github.com/tomwright/dasel/v2.BasicFunction.Run(...)
	github.com/tomwright/dasel/v2/func.go:144
github.com/tomwright/dasel/v2.(*Step).execute(0xc000313810)
	github.com/tomwright/dasel/v2/step.go:30 +0xc3
github.com/tomwright/dasel/v2.(*Context).Next(0xc03530de00)
	github.com/tomwright/dasel/v2/context.go:227 +0x1a5
github.com/tomwright/dasel/v2.(*Context).Run(0x7aa6e0?)
	github.com/tomwright/dasel/v2/context.go:195 +0x29
github.com/tomwright/dasel/v2.(*Context).subSelect(0xc0322b8d80, {0x7aa6e0?, 0xc0285af980?}, {0xc0322e65e0?, 0x2?})
	github.com/tomwright/dasel/v2/context.go:162 +0x67
github.com/tomwright/dasel/v2.glob..func6.1({{0xc0322d2340, 0xc028351f80, 0x16}, 0xc028706380, 0x0, 0xc0285af860}, {{0xc0322e65e0, 0xc}, {0xc0322e65f0, 0xe}})
	github.com/tomwright/dasel/v2/func_equal.go:42 +0x85
github.com/tomwright/dasel/v2.glob..func6(0xc0322b8d80, 0xc0003135e0, {0xc032232da0?, 0x2, 0xa61cc0?})
	github.com/tomwright/dasel/v2/func_equal.go:64 +0x706
github.com/tomwright/dasel/v2.BasicFunction.Run(...)
	github.com/tomwright/dasel/v2/func.go:144
github.com/tomwright/dasel/v2.(*Step).execute(0xc0003135e0)
	github.com/tomwright/dasel/v2/step.go:30 +0xc3
github.com/tomwright/dasel/v2.(*Context).Next(0xc0322b8d80)
	github.com/tomwright/dasel/v2/context.go:227 +0x1a5
github.com/tomwright/dasel/v2.(*Context).Run(0x7aa6e0?)
	github.com/tomwright/dasel/v2/context.go:195 +0x29
github.com/tomwright/dasel/v2.Select({0x7aa6e0?, 0xc0285af890?}, {0x7ffce70d5591?, 0xc000138840?})
	github.com/tomwright/dasel/v2/context.go:117 +0x2a
github.com/tomwright/dasel/v2/internal/command.runSelectCommand(0xc0000c1c28, 0x7b3e8f?)
	github.com/tomwright/dasel/v2/internal/command/select.go:99 +0xbc
github.com/tomwright/dasel/v2/internal/command.selectRunE(0xc00019ac00?, {0xc00035bd80, 0x1, 0x7ad020?})
	github.com/tomwright/dasel/v2/internal/command/select.go:83 +0x4f9
github.com/spf13/cobra.(*Command).execute(0xc000004300, {0xc0000160a0, 0x8, 0x8})
	github.com/spf13/cobra@v1.8.0/command.go:983 +0xabc
github.com/spf13/cobra.(*Command).ExecuteC(0xc000004300)
	github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).Execute(0x408e2b?)
	github.com/spf13/cobra@v1.8.0/command.go:1039 +0x13
main.main()
	github.com/tomwright/dasel/v2/cmd/dasel/main.go:10 +0x25
```

</details>

**To Reproduce**
Steps to reproduce the behavior:

1. Create the following CSV (or any other, I think):
    ```csv
    Hello;There;
    1;2;
    ```
2. Run `dasel -f foo.csv --csv-comma ';' -w json --pretty  'equal([], )'`
3. Observe the error

I have to admit I didn't look much at the dasel query syntax (this is a reduced version), I just ended up piping it to `jq` and using some `select` and `map` magic for my usecase :)

**Expected behavior**
It should not crash.

**Desktop:**

- OS: Arch Linux (using the nix package though)
- Version `dasel version 2.6.0`

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,7 +6,14 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+
 - Nothing yet.
+
+## [v2.7.1] - 2024-03-14
+
+### Fixed
+
+- Fixed a bug that could cause a panic. 
 
 ## [v2.7.0] - 2024-03-14
 
@@ -664,7 +671,8 @@ See [documentation](https://daseldocs.tomwright.me) for all changes.
 
 - Everything!
 
-[unreleased]: https://github.com/TomWright/dasel/compare/v2.7.0...HEAD
+[unreleased]: https://github.com/TomWright/dasel/compare/v2.7.1...HEAD
+[v2.7.0]: https://github.com/TomWright/dasel/compare/v2.7.0...v2.7.1
 [v2.7.0]: https://github.com/TomWright/dasel/compare/v2.6.0...v2.7.0
 [v2.6.0]: https://github.com/TomWright/dasel/compare/v2.5.0...v2.6.0
 [v2.5.0]: https://github.com/TomWright/dasel/compare/v2.4.1...v2.5.0

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,10 +7,6 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
-- Nothing yet.
-
-## [v2.7.1] - 2024-03-14
-
 ### Fixed
 
 - Fixed a bug that could cause a panic. 
@@ -671,8 +667,7 @@ See [documentation](https://daseldocs.tomwright.me) for all changes.
 
 - Everything!
 
-[unreleased]: https://github.com/TomWright/dasel/compare/v2.7.1...HEAD
-[v2.7.0]: https://github.com/TomWright/dasel/compare/v2.7.0...v2.7.1
+[unreleased]: https://github.com/TomWright/dasel/compare/v2.7.0...HEAD
 [v2.7.0]: https://github.com/TomWright/dasel/compare/v2.6.0...v2.7.0
 [v2.6.0]: https://github.com/TomWright/dasel/compare/v2.5.0...v2.6.0
 [v2.5.0]: https://github.com/TomWright/dasel/compare/v2.4.1...v2.5.0

--- a/internal/command/select_test.go
+++ b/internal/command/select_test.go
@@ -441,4 +441,13 @@ d.e.f`)),
 		}
 	})
 
+	t.Run("Issue 392 panic", runTest(
+		[]string{"-r", "csv", "--csv-comma", ";", "-w", "json", "equal([], )"},
+		[]byte(`Hello;There;
+1;2;`),
+		[]byte("false\n"),
+		nil,
+		nil,
+	))
+
 }

--- a/value.go
+++ b/value.go
@@ -1,9 +1,10 @@
 package dasel
 
 import (
+	"reflect"
+
 	"github.com/tomwright/dasel/v2/dencoding"
 	"github.com/tomwright/dasel/v2/util"
-	"reflect"
 )
 
 // Value is a wrapper around reflect.Value that adds some handy helper funcs.
@@ -283,7 +284,8 @@ func (v Value) Index(i int) Value {
 func (v Value) Append() Value {
 	currentLen := v.Len()
 	newLen := currentLen + 1
-	updatedSlice := reflect.MakeSlice(sliceInterfaceType, newLen, newLen)
+
+	updatedSlice := reflect.MakeSlice(reflect.TypeOf(v.Interface()), newLen, newLen)
 	// copy all existing elements into updatedSlice.
 	// this leaves the last element empty.
 	for i := 0; i < currentLen; i++ {
@@ -292,18 +294,21 @@ func (v Value) Append() Value {
 		)
 	}
 
-	v.FirstAddressable().Set(updatedSlice)
+	firstAddressable := v.FirstAddressable()
+	firstAddressable.Set(updatedSlice)
 
+	// This code was causing a panic...
+	// I'm not sure if it's necessary - all tests continue to pass.
 	// Set the last element to uninitialised.
-	updatedSlice.Index(currentLen).Set(
-		v.Index(currentLen).asUninitialised().Value,
-	)
+	//updatedSlice.Index(currentLen).Set(
+	//	v.Index(currentLen).asUninitialised().Value,
+	//)
 
 	return v
 }
 
-var sliceInterfaceType = reflect.TypeOf([]interface{}{})
-var mapStringInterfaceType = reflect.TypeOf(map[string]interface{}{})
+var sliceInterfaceType = reflect.TypeFor[[]any]()
+var mapStringInterfaceType = reflect.TypeFor[map[string]interface{}]()
 
 var UninitialisedPlaceholder interface{} = "__dasel_not_found__"
 

--- a/value.go
+++ b/value.go
@@ -316,7 +316,8 @@ func (v Value) Append() Value {
 	firstAddressable.Set(updatedSlice)
 
 	// This code was causing a panic...
-	// I'm not sure if it's necessary - all tests continue to pass.
+	// It doesn't seem necessary. Leaving here for reference in-case it was needed.
+	// See https://github.com/TomWright/dasel/issues/392
 	// Set the last element to uninitialised.
 	//updatedSlice.Index(currentLen).Set(
 	//	v.Index(currentLen).asUninitialised().Value,

